core/lib/src/server.rs

```diff
@@ -64,23 +64,58 @@ async fn handle<Fut, T, F>(name: Option<&str>, run: F) -> Option<T>
 async fn hyper_service_fn(
     rocket: Arc<Rocket<Orbit>>,
     conn: ConnectionMeta,
-    hyp_req: hyper::Request<hyper::Body>,
+    mut hyp_req: hyper::Request<hyper::Body>,
 ) -> Result<hyper::Response<hyper::Body>, io::Error> {
     // This future must return a hyper::Response, but the response body might
     // borrow from the request. Instead, write the body in another future that
     // sends the response metadata (and a body channel) prior.
     let (tx, rx) = oneshot::channel();
 
     tokio::spawn(async move {
+        // Upgrade before do any other; we handle errors below
+        let hyp_upgraded = hyper::upgrade::on(&mut hyp_req);
+
         // Convert a Hyper request into a Rocket request.
         let (h_parts, mut h_body) = hyp_req.into_parts();
         match Request::from_hyp(&rocket, &h_parts, Some(conn)) {
             Ok(mut req) => {
                 // Convert into Rocket `Data`, dispatch request, write response.
                 let mut data = Data::from(&mut h_body);
                 let token = rocket.preprocess_request(&mut req, &mut data).await;
-                let response = rocket.dispatch(token, &mut req, data).await;
-                rocket.send_response(response, tx).await;
+                let mut response = rocket.dispatch(token, &req, data).await;
+
+                if response.status() == Status::SwitchingProtocols {
+                    let may_upgrade = response.upgrade_mut().take();
```

Comment:
Hi and thanks for the feedback!

`upgrade_mut().take()` is needed here because inside the match `send_response(response,tx)` is called, which we need so the `hyper::Upgrade` struct is actually valid, but it dosnt take a reference for the reponse and as such moves the response out of scope. With `take()`, the `Upgrade` struct is taken out of the response and the code can own it as long as it needs it.

Sure, one possiblity might be here to change `send_response()` so it takes a ref. Tested it and it always errors with that the async block in `hyper_service_fn()` is not `Send`:
```
error: future cannot be sent between threads safely
   --> /home/mai/projects/github/Rocket/core/lib/src/server.rs:74:18
    |
74  |       tokio::spawn(async move {
    |  __________________^
75  | |         // Upgrade before do any other; we handle errors below
76  | |         let hyp_upgraded = hyper::upgrade::on(&mut hyp_req);
77  | |
...   |
132 | |         }
133 | |     });
    | |_____^ future created by async block is not `Send`
    |
    = help: the trait `Sync` is not implemented for `dyn tokio::io::AsyncRead + std::marker::Send`
note: future is not `Send` as this value is used across an await
   --> /home/mai/projects/github/Rocket/core/lib/src/server.rs:154:49
    |
144 |         response: &Response<'_>,
    |         -------- has type `&response::response::Response<'_>` which is not `Send`
...
154 |         match self._send_response(&response, tx).await {
    |                                                 ^^^^^^ await occurs here, with `response` maybe used later
...
159 |     }
    |     - `response` is later dropped here
note: required by a bound in `tokio::spawn`
   --> /home/mai/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/task/spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ^^^^ required by this bound in `spawn`
```

Reply:
With a bit of refactoring, it's possible to don't use `take` nor having to expose a mutable getter `upgrade_mut`.

```rust
if response.upgrade().is_some() {
    // send the finishing response; needed so that hyper can upgrade the request
    rocket.send_response(response, tx).await;

    let upgrade = response.upgrade().unwrap();

    /* rest of the code */
} else {
    error_!("Status is 101 switching protocols, but response dosn't hold a upgrade");
    let response = rocket.handle_error(Status::InternalServerError, &req).await;
    rocket.send_response(response, tx).await;
}
```

Reply:
Sadly when adding your propasal, the same error but at different place arises:

<details>
    <summary>Error log</summary>

```
error: future cannot be sent between threads safely
   --> /home/mai/projects/github/Rocket/core/lib/src/server.rs:74:18
    |
74  |       tokio::spawn(async move {
    |  __________________^
75  | |         // Upgrade before do any other; we handle errors below
76  | |         let hyp_upgraded = hyper::upgrade::on(&mut hyp_req);
77  | |
...   |
155 | |         }
156 | |     });
    | |_____^ future created by async block is not `Send`
    |
    = help: the trait `Sync` is not implemented for `dyn Upgrade<'_> + std::marker::Send`
note: future is not `Send` as this value is used across an await
   --> /home/mai/projects/github/Rocket/core/lib/src/server.rs:95:43
    |
93  |                         let upgrade = response.upgrade().unwrap();
    |                             ------- has type `&Box<dyn Upgrade<'_> + std::marker::Send>` which is not `Send`
94  |
95  |                         match hyp_upgraded.await {
    |                                           ^^^^^^ await occurs here, with `upgrade` maybe used later
...
109 |                     } else {
    |                     - `upgrade` is later dropped here
note: required by a bound in `tokio::spawn`
   --> /home/mai/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.21.2/src/task/spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ^^^^ required by this bound in `spawn`
```
</details>

When commenting out the whole `match hyp_upgraded.await` it becomes clear, that an borrow of moved value error is the real culprit:
```
error[E0382]: borrow of moved value: `response`
  --> /home/mai/projects/github/Rocket/core/lib/src/server.rs:93:39
   |
85 |                 let mut response = rocket.dispatch(token, &req, data).await;
   |                     ------------ move occurs because `response` has type `response::response::Response<'_>`, which does not implement the `Copy` trait
...
91 |                         rocket.send_response(response, tx).await;
   |                                              -------- value moved here
92 |
93 |                         let upgrade = response.upgrade().unwrap();
   |                                       ^^^^^^^^^^^^^^^^^^ value borrowed here after move
```

Whats so bad with `take()` anyway? I mean `send_response()` certainly dosnt need the `Upgrade` and afterwards the code has the ownage about it so it's also get correctly dropped.

Reply:
Oops I forgot the `await`. You could try `response.upgrade().unwrap().start(hyp_upgraded)`.

> Out of curiosity: Whats so bad with take() anyway? I mean send_response() certainly dosnt need the Upgrade and afterwards the code has the ownage about it so it's also get correctly dropped.

Using `take()` introduce API that alter the response were it could be avoided.

Reply:
Using `response.upgrade().unwrap().start(hyp_upgraded)` which creates the following code, it still dosnt compile:
```rust
let mut response = rocket.dispatch(token, &req, data).await;

if response.status() == Status::SwitchingProtocols {

if response.upgrade().is_some() {
    // send the finishing response; needed so that hyper can upgrade the request
    rocket.send_response(response, tx).await;

    match hyp_upgraded.await {
        Ok(hyp_upgraded) => {
            // let the upgrade take the upgraded hyper request
            let fu = response.upgrade().unwrap().start(hyp_upgraded);
            fu.await;
        }
        Err(e) => {
            error_!("Failed to upgrade request: {e}");
            // NOTE: we *should* send a response here but since we send one earlier AND upgraded the request,
            //       this cannot be done easily at this point...
            // let response = rocket.handle_error(Status::InternalServerError, &req).await;
            // rocket.send_response(response, tx).await;
        }
    }
} else {
    error_!("Status is 101 switching protocols, but response dosn't hold a upgrade");
    let response = rocket.handle_error(Status::InternalServerError, &req).await;
    rocket.send_response(response, tx).await;
}
```

Error:
```
error[E0382]: borrow of moved value: `response`
  --> /home/mai/projects/github/Rocket/core/lib/src/server.rs:96:42
   |
85 |                 let response = rocket.dispatch(token, &req, data).await;
   |                     -------- move occurs because `response` has type `response::response::Response<'_>`, which does not implement the `Copy` trait
...
91 |                         rocket.send_response(response, tx).await;
   |                                              -------- value moved here
...
96 |                                 let fu = response.upgrade().unwrap().start(hyp_upgraded);
   |                                          ^^^^^^^^^^^^^^^^^^ value borrowed here after move

For more information about this error, try `rustc --explain E0382`.
```

I tink the only possibility is either to move / take out the upgrade out of the response or to rework `send_response` to borrow the response instead of moving it.

Reply:
The option of using `take_upgrade` also seems best to me. Added a commit.